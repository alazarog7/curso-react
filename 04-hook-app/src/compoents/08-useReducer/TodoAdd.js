import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({
    handleAddTodo
}) => {

    const [ {description}, handleInputChange, reset] = useForm({description: ''})

    const handleSubmit = (e) => {
        e.preventDefault();
        if(description !=  ''){
            handleAddTodo({
                id: new Date().getTime(),
                desc: description,
                done: false
            });
        }

        reset();
    }


    return (
        <form onSubmit={ handleSubmit }>
            <input type="text" 
                    name="description"
                    className="form-control"
                    placeholder="Aprender ..."
                    value={ description }
                    onChange={handleInputChange}
                    autoComplete="off"/>
            <div className="d-grid gap-2">
                <button type="submit"
                        className="btn btn-outline-primary mt-1">
                        Agregar
                </button>
            </div>
        </form>
    )
}
