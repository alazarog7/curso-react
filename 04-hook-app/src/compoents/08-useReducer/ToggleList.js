import React from 'react'
import { ToggleListItem } from './ToggleListItem'

export const ToggleList = ({
    todos,
    handleDelete,
    handleToggle
}) => {
    return (
          <ul className="list-group">
              {
                  todos.map( (t, index) => <ToggleListItem key={ t.id } todo={t} index={index} handleDelete={handleDelete} handleToggle={handleToggle}/>)
              }
           </ul>  
        
    )
}
