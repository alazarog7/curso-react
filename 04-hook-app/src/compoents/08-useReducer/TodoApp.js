import React, { useEffect, useReducer } from 'react'
import './style.css'
import todoReducer from './TodoReducer';
import { ToggleList } from './ToggleList';
import { TodoAdd } from './TodoAdd';


const init = () => {
    //localStorage.setItem('todos',[{
    //    id: new Date().getTime(),
    //    desc: 'Aprender React',
    //    done: false
    //}]);
    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {
    const [ todos, dispatch ] = useReducer(todoReducer, [], init)

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (id) => {
        
        dispatch({
            type:'delete',
            payload:id
        });
    }

    const handleAddTodo = ( newTodo ) => {

        dispatch(  {
            type:'add',
            payload: newTodo
        } );

    }

    const handleToggle = ( todoId ) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        })
    }

    return (
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr/>

            <div className="row">
                <div className="col-5">
                    <ToggleList todos={todos} handleDelete={handleDelete} handleToggle={handleToggle}/>
                </div>
                <div className="col-7">
                    <TodoAdd handleAddTodo={handleAddTodo} />
                </div>
            </div>    

            
        </div>
    )
}
