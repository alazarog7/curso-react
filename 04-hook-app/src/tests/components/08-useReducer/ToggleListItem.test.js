import { shallow } from "enzyme"
import { ToggleListItem } from "../../../compoents/08-useReducer/ToggleListItem"
import { demoTodos } from "../../fixtures/demoTodo";

describe('TodoListItem', () => {
    
    let wrapper = null;
    const handleToggle = jest.fn();
    const handleDelete = jest.fn();
    
    beforeEach(()=>{
        jest.clearAllMocks();
        wrapper = shallow(<ToggleListItem
                                todo={demoTodos[0]}
                                index={1}
                                handleToggle={handleToggle}
                                handleDelete={handleDelete}
                          />)
    })

    afterEach(()=>{
        wrapper = null
    })

    test('should show component correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('should call handleDelete function', () => {
       wrapper.find('button').simulate('click');
       expect(handleDelete).toHaveBeenCalled();
    });

    test('should call handleToggle function', () => {
       wrapper.find('p').simulate('click');
       expect(handleToggle).toHaveBeenCalledWith(expect.any(Number));
    });

    test('should return the correct message',()=>{
        expect(wrapper.find('p').text().trim()).toBe(
            '2. jugar'
        )
    });

    test('should have complete class', () => {
        
        demoTodos[0].done = true;
        wrapper = shallow(<ToggleListItem
            todo={demoTodos[0]}
            index={1}
            handleToggle={handleToggle}
            handleDelete={handleDelete}
        />)
        expect(wrapper.exists('.complete')).toEqual(true);
    })
})
