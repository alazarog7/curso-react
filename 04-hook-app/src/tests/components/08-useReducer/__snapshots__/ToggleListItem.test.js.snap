// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem should show component correctly 1`] = `
<li
  className="list-group-item"
>
  <p
    className="text-center false"
    onClick={[Function]}
  >
     
    2
    . 
    jugar
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
